### ISSUE TITLE
Deprecate astropy test runner in core and downstream

### ISSUE THREAD

**Issue by pllim** [astropy/astropy#16177]
This is what some people call "an epic" in JIRA-land.

A long, long time ago, in the same galaxy, `astropy` existed when Python was youngish, pytest was younger (remember when we bundled pytest as a single file?) , CI on GitHub was almost non-existent (ask Tom R about Jenkins). There was a need to be able to test installed versions of `astropy`. And thus, the test runner was created. For many years, it served us well.

https://github.com/astropy/astropy/blob/52d2b360dc386c87e4e7a74dfdacd28428a23a30/astropy/tests/runner.py#L290-L293

https://github.com/astropy/astropy/blob/52d2b360dc386c87e4e7a74dfdacd28428a23a30/astropy/__init__.py#L137

We even advertised it in `package-template` (https://github.com/astropy/package-template/blob/36ed48db74c9a0bfcf214e14a1a3fb1fdaae0c6c/%7B%7B%20cookiecutter.package_name%20%7D%7D/%7B%7B%20cookiecutter.module_name%20%7D%7D/_%7B%7B%20cookiecutter._parent_project%20%7D%7D_init.py#L13), so downstream packages may have similar runners too (whether they realized it is there or not is another issue).

Now in 2024, we have multi-architecture CI and wheels. The need for such a test runner has diminished greatly. In fact, having it might even hinder progress like these:

* https://github.com/astropy/astropy/issues/16152
* https://github.com/astropy/astropy/issues/13678

Additionally, it is not aging well now that pytest is very complicated and there are plenty of pytest plugins available:

* https://github.com/astropy/astropy/issues/16165

All the while, it was never feature complete:

* #10667
* https://github.com/astropy/astropy/issues/8153 

The plea to get rid of the test runner went as far back as 2018 (https://github.com/astropy/astropy/issues/8153#issuecomment-439667910), *"Friends, I beg of you, let's please consider axing the test runner entirely."*

# Proposal

1. Deprecate `astropy.test()` function and `astropy.tests.runner` module.
2. Open issues in affected downstream packages to warn them to check whether they inherited this test runner from package template (or added it intentionally) and tell them to move away from using it. Alternative is to invoke `pytest` directly (or `tox`, whatever).
3. After a deprecation period, remove the deprecated stuff from `astropy`. If downstream packages have not caught up, they need to pin to an older release of `astropy` or have a broken test runner.

Impact: High!
Benefit: Less technical debt to carry forward.
Cost: Some crying maintainers.

# Counter proposal

We keep everything as-is, but we have to start testing `astropy.test()` in our CI.

**Comment by namurphy** [comment]
I did a [search for "astropy.test()" in Python files across GitHub](https://github.com/search?q=%22astropy.test%28%29%22+path%3A*.py&type=code&ref=advsearch) and got 76 results. A similar [search for "astropy.tests.runner" in Python files across GitHub](https://github.com/search?q=%22astropy.tests.runner%22+path%3A*.py&type=code&ref=advsearch) got 132 results.  (Checking on this was waaaay more fun than timing Sphinx builds.)

**Comment by MridulS** [comment]
Thanks for the write up @pllim! I wonder who is the audience for the deprecation? Can we remove `astropy.test()` directly and only deprecate `TestRunner`? Assuming the hypothesis that an end user doesn't really need to run `astropy.test()` today.

There are a bunch of packages which use the `TestRunner` code downstream but from a quick glance it seems like there isn't much magic happening in the downstream packages, it just directly imports `TestRunner` and then exposes a test function in the package namespace, so the change hopefully wouldn't be that painful!

>I did a [search for "astropy.test()" in Python files across GitHub](https://github.com/search?q=%22astropy.test%28%29%22+path%3A*.py&type=code&ref=advsearch) and got 76 results.

If we remove all the forks on github, the only active place `astropy.test()` seemed to be used is in https://github.com/spack/spack/blob/31de670bd26beca979ebd75dcb0ce90c535a78c4/var/spack/repos/builtin/packages/py-astropy/package.py#L126 and they do support running tests directly with `pytest`.

**Comment by pllim** [comment]
Thanks for the searches, y'all!

We advertise `astropy.test()` in every RC testing (e.g., https://github.com/astropy/astropy/wiki/v6.0-RC-Testing), so I think deprecation is safer bet than straight removal. But whether `astropy.test()` is considered "public API" or not perhaps could be debated, especially since the public API APE is still WIP:

* https://github.com/astropy/astropy-APEs/pull/85

**Comment by pllim** [comment]
Here is an example of active downstream usage of the test runner:

https://github.com/spacetelescope/synphot_refactor/blob/c60d653f311c552b6256099dfc057d3f0a44d176/.github/workflows/publish-to-pypi.yml#L46

**Comment by mhvk** [comment]
I agree we can stop using the test runner (and deprecate the `TestRunner` class). The main thing really would seem to have *good* instructions for how people can test astropy without it. The nice thing about our current test instructions at https://github.com/astropy/astropy/wiki/v6.0-RC-Testing is that we do not have to worry at all about what virtual environment people have, etc. But I *think* one can replace the second line with just a `tox` incantation, no?

**Comment by saimn** [comment]
>  The nice thing about our current test instructions at https://github.com/astropy/astropy/wiki/v6.0-RC-Testing is that we do not have to worry at all about what virtual environment people have, etc. But I think one can replace the second line with just a tox incantation, no?

As long as tests are installed, one can use `pytest --pyargs astropy`.
Should be quite similar to `astropy.test()` without a local clone.

**Comment by saimn** [comment]
A search for [`make_test_runner_in` without forks](https://github.com/search?q=%22make_test_runner_in%22+path%3A*.py+NOT+is%3Afork+&type=code&ref=advsearch) gives 84 results, seems to give a good list of packages importing the TestRunner (i.e. using the template).

**Comment by pllim** [comment]
> As long as tests are installed

But how, if the wheels no longer come with tests?

**Comment by mhvk** [comment]
Presumably `pip install astropy[test]` would pull in `astropy-tests` too!

**Comment by pllim** [comment]
Oh, we talking about moving the tests out of this repo too? 😱

**Comment by Cadair** [comment]
SunPy :fire:'ed the test runner a long time ago, it's very easy to just call `pytest.main` and get the same effect: https://github.com/sunpy/sunpy/blob/main/sunpy/tests/self_test.py#L73

**Comment by bsipocz** [comment]
We do actually use `astropy.test()` in astropy's conda recipe 🤷‍♀️

**Comment by pllim** [comment]
Update: astropy conda recipe no longer uses `astropy.test()` and I have #17732 to stop using it in our "double run" CI.

**Comment by pllim** [comment]
I also updated my proposal section above so it makes more sense (at least to me).

**Comment by pllim** [comment]
As far as my search-fu allows, 194 repos at least import TestRunner. Whether it is actually used or not is another story; but the impact is definitely non-zero. I can warn them with batch issues.

**Comment by keflavich** [comment]
What percent of the crying maintainers are me?  😢   Looks like about 5%.

Anyway, thanks for this @pllim.   Can you remind me where to find instructions on how best to move away from using astropy for testing?  I'm pretty sure this all comes from inheriting from the pre-cookiecutter package template.

**Comment by mhvk** [comment]
Only 1.5% for me... But, yes, from the same template and will now have to update my user-facing instructions about how to test.... But it is good to get rid of old cruft...

**Comment by weaverba137** [comment]
I don't have much to do in my package fortunately. I already use pytest/tox; the test runner is just leftover inheritance from package-template, but a pointer to recent documentation of replacements would be appreciated, specifically running tests on a package that is already installed.

**Comment by pllim** [comment]
😅 Not sure how much this helps but this kinda shows a little example of how to replace specific `test()` code with `pytest` directly:

* https://github.com/conda-forge/astropy-feedstock/pull/151
* #17732

My own packages are also affected. When I get around to them, I can link my PRs here as well.

**Comment by pllim** [comment]
Examples so far... they vary in amount of diff for various reasons.

* https://github.com/spacetelescope/jdaviz/pull/3451
* https://github.com/spacetelescope/lcviz/pull/172
* https://github.com/astropy/specutils/pull/1214

**Comment by pllim** [comment]
Update: I put in https://github.com/astropy/astropy/pull/17874 to kick off a pending deprecation period. 😬 🔥

**Comment by pllim** [comment]
OK actual deprecation is happening for v8.0. Please review https://github.com/astropy/astropy/pull/17883 . Thanks!
